Add Checkbox component

obulat opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #405 by @obulat

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR creates a standalone `Checkbox` component with the new styles, and uses it throughout the app:
<img width="330" alt="Screen Shot 2021-11-09 at 5 53 12 PM" src="https://user-images.githubusercontent.com/15233243/140947040-646fd39b-1bf7-41e9-9ee6-9348b4101b08.png">
The last checkbox in the screenshot is disabled. I couldn't find the actual disabled variant in Figma, @panchovm .

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
Checkout the branch and run `npm run storybook` to see the Checkbox story there, and verify that it works both with a string label, and a `License` component label.
Then, run `npm run dev` and test the Checkboxes in use on the home page (license types inputs) and the filter sidebar. Check/uncheck them to see that the filters are applied, and new media is fetched.
`npm run test` should also run without problems :)
## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

obulat commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,122 @@
+
+.checkbox:checked::after {
+  content: '';
+  @apply w-2 h-4 border-2 border-white absolute top-0 left-1/3;
```
```suggestion
  @apply w-2 h-4 border border-white absolute top-0 left-1/3;
```

obulat commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,122 @@
+.checkbox:checked::after {
+  content: '';
+  @apply w-2 h-4 border-2 border-white absolute top-0 left-1/3;
+  border-width: 1px;
+  border-top: 0;
```
```suggestion
  border-top: 0;
```

sarayourfriend commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,122 @@
+      type: String,
+      required: true,
+    },
+    /**
+     * The checkbox can be labeled either using the `label` prop, or the default
+     * slot (for labels that can contain more than simple strings).
+     *
+     * @default ''
+     */
+    label: {
+      type: String,
+      default: '',
+    },
```
Having two ways to set the label seems confusing to me and also potentially prone to someone forgetting to add a label. What if we go with just the slot and somehow verify that it is passed? Is there a scenario where we wouldn't be able to pass the slot value and would have to use a prop instead?

obulat replied:
I've removed the label prop. I guess it was my first thought to add the label using prop, and then I just went with it without much thought. AND got confused with whether I need to pass a prop or not myself, so thank you for this comment!

dhruvkb commented:
The padding on the bottom and right sides seems smaller than the other two.

sarayourfriend commented on test/unit/specs/components/checkbox.spec.js:
```diff
@@ -0,0 +1,31 @@
+      slots: { default: ByIcon },
+    })
+    const checkbox = screen.getByRole('checkbox')
+    await checkbox.click()
```
Firing events like this work most of the time, but it's recommended to use `userEvent.click` instead of calling `click` on the element directly. Calling `click` on the element only fires the `click` event, whereas an actual click using a mouse device will cause a `mousedown, click, mouseup` series of events. `userEvent` simulates this kind of interaction.
```suggestion
    await userEvent.click(checkbox)
```

sarayourfriend commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,121 @@
+  },
+}
+</script>
+<style>
```
We should use `scoped` or `module` to avoid introducing global classes and/or clashing with existing global classes.

sarayourfriend commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,121 @@
+
+<script>
+export default {
+  name: 'Checkbox',
```
Single word component names are not recommended in Vue because you could inadvertently override some future HTML native element by the same name (all native elements have a single word name, even `textarea` which is two words in the prototype name `HTMLTextAreaElement` 😱)

For that reason I was appending `V` to `VButton` for example. This helps future proof the Vue components (even though it makes the names less pleasant 😞)

sarayourfriend commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,121 @@
+      :value="inputValue"
+      :disabled="disabled"
+      @change="onChange"
+    /><span>
+      <slot>
+        {{ label }}
+      </slot>
+    </span>
```
I get why prettier auto-formats things this way because of Vue's whitespace handling, but in this case the whitespace means nothing, so for readability's sake I think it would be better to put each tag on a separate line.

```suggestion
    />
    <span>
      <slot>
        {{ label }}
      </slot>
    </span>
```

sarayourfriend commented on src/components/Filters/FilterChecklist.vue:
```diff
@@ -34,35 +34,33 @@
-          {{ itemLabel(item) }}
-        </label>
-        <button class="appearance-none" type="button">
+          {{ itemLabel(item) }}</Checkbox
+        >
```
Same thing here with the whitespace, it doesn't affect the output so we can put the Checkbox closing tag all on one line to make it easier to find compared to the opening tag.

```suggestion
          {{ itemLabel(item) }}
        </Checkbox>
```

sarayourfriend commented on test/unit/specs/components/checkbox.spec.js:
```diff
@@ -0,0 +1,31 @@
+  })
+
+  it('should render icons if provided in the default slot', async () => {
+    // @obulat: I cannot pass a required prop to `License` object to use it in the slot,
+    // nor can I check that svg is rendered using its accessible role. But if you use
+    // `screen.debug()`, you can see it is rendered :)
```
Can we just use `querySelector` for `svg` on the container that `render` returns?

```js
const { container } = render(Checkbox, ...)

const icon = container.querySelector('svg')
```

Also for being able to render a full output with slots with props, I think it's appropriate to create a TestWrapper style component:

```js
const TestWrapper = Vue.component('TestWrapper', {
  template: `<Checkbox id="simple" label="Is Code Poetry?"><Licence :whatever="'foo'" /></Checkbox>`,
  components: { Checkbox, License },
})
```

That way you can render anything and everything that a regular Vue application would allow you to without having to juggle around the slots options that the `render` function gives you and its limitations.

obulat replied:
Thank you so much for this, @sarayourfriend ! I haven't seen `container` yet, and the `TestWrapper` seems just the solution I was looking for!

sarayourfriend commented on test/unit/specs/components/checkbox.spec.js:
```diff
@@ -0,0 +1,31 @@
+  it('should render a checkbox with a string label', () => {
+    render(Checkbox, { propsData: { id: 'simple', label: 'Is Code Poetry?' } })
+    // Get throws an error if element is not found, so this serves as an assert
+    screen.getByLabelText(/code/i, { checked: false })
```
While these do throw errors if the element isn't found, I think for consistency and clarity it would still be nice to add an explicit assertion.

```suggestion
    const checkbox = screen.getByLabelText(/code/i)
    expect(checkbox).toHaveAttribute('checked', 'false')
```

For example

sarayourfriend commented on src/components/HeroSection.vue:
```diff
@@ -92,15 +94,24 @@ export default {
       this.setSearchTerm({ q: this.form.searchTerm })
+      Object.keys(this.filters)
+        .filter((filterName) => !!this.filters[filterName])
+        .map((f) => this.checkFilter({ filterType: 'licenseTypes', code: f }))
```
This can just be a `forEach` instead of a `map` I think. Might save some tiny amount of CPU cycles by not throwing away the `map` result 😅 Also it's more semantically obvious what is meant... usually with a `map` or `reduce` operation I expect the result to be used, whereas `forEach` is explicitly just a loop over the array and nothing more.

obulat replied:
Thank you for your explanation :)

sarayourfriend commented on src/components/Checkbox.vue:
```diff
@@ -0,0 +1,121 @@
+      :value="inputValue"
+      :disabled="disabled"
+      @change="onChange"
+    /><span>
```
Is the `span` wrapper necessary for the label text? I removed it locally and it didn't seem to effect the visual output at all.

obulat replied:
It did change the visual output for me in the beginning, but that might be because I hadn't deleted all the sass styles before, I will check tomorrow.

sarayourfriend reviewed (commented):
LGTM so far! I love the implementation, sweet and simple 🙂 Just a few thoughts of things we could try differently and some small things to fix here and there. But overall this is awesome 😀

panchovm commented:
The color for `active` state looks good, but the check icon within the box looks bigger and misaligned. The component is based on the WordPress Design Library version. [Here is the storybook](https://wordpress.github.io/gutenberg/?path=/story/components-checkboxcontrol--default).

I noticed there is no style for the `disabled` state in Figma either. But my guess is there is one in the code version of the component.

zackkrida commented:
Related to @panchovm's reference to Gutenberg, I think we'll want to take their approach and use `appearance: none;` on the actual checkbox while positioning an svg over the real element. Just inspecting their implementation in the browser should be really useful for reviewing. Thanks for sharing that @panchovm

zackkrida commented:
There's a regression where in the checked state I don't even see a checkmark anymore:

<img width="950" alt="CleanShot 2021-11-10 at 15 33 25@2x" src="https://user-images.githubusercontent.com/6351754/141189122-46bd3961-6fad-47e7-a442-034d32c9408a.png">

obulat commented:
> There's a regression where in the checked state I don't even see a checkmark anymore:
> 
The first implementation kept the Vue `checked` state from the parent, and did not have any local state. This is why on click the checkbox input was checked in HTML, and used the `.checkbox:checked` styles (i.e. black background). But the JavaScript `checked` value wasn't update by the parent, and therefore stayed `false`, which prevented the checkmark from showing.

I've rewritten the component, adding local checked state, so that it can be in sync with the HTML `checked` attribute, but can also be updated by the parent. I also used composition API, and removed the `value` prop from the component as it doesn't really add value, and collides with the `ref.value` values.

krysal reviewed (commented):
The component looks identical to the Figma mockup, great! I just wonder if we can add a `value` prop for the _HTML value attribute_. The licenses filter is a perfect example for when the `name` is the name of a field (`licenses`) and the value is its possible values, worth the redundancy, like licenses codes in this case, and they might not be unique so using the `id` always might not be the desire in those cases.

zackkrida commented on src/components/meta/Checkbox.stories.js:
```diff
@@ -0,0 +1,75 @@
+import VCheckbox from '~/components/VCheckbox'
+import License from '~/components/License/License'
```
Nice detail testing the Storybook _build_ @krysal, thank you!

dhruvkb commented on src/components/VCheckbox.vue:
```diff
@@ -0,0 +1,133 @@
+import { computed, defineComponent, ref, watch } from '@nuxtjs/composition-api'
+
+const VCheckbox = defineComponent({
+  name: 'VCheckbox',
```
Might be nice to set [the `model` option](https://vuejs.org/v2/api/#model) to the component to use with `v-model`.

obulat replied:
I wanted to use the `v-model` at first, too. The problem with model for checkbox is it's binding `value` and `input` event. And in checkbox, `value` is the name of the checkbox value (ie., 'CC BY' for licenses), and the thing that actually needs to change is the `checked` property.

dhruvkb commented on src/components/VCheckbox.vue:
```diff
@@ -0,0 +1,133 @@
+     * the checkbox. id is also used in the `change` event payload as the `name`
+     * and `value` parameters if they are not set.
+     */
+    id: {
```
I'm not 100% sure but since ID is an HTML attribute, this might conflict. For disambiguation, I used `fieldId` in the `InputField` component.

obulat replied:
This `id` prop is used both for the HTML attribute (mainly to get the `<label for="id">` working well, since I've read that simply wrapping an input in a label might not be sufficient for some screen readers), and to use as a fall-back for `name` and/or `value` properties. I can see that a non-unique id might cause a problem, but don't know if it can be significant. Can it?

dhruvkb commented on src/components/VCheckbox.vue:
```diff
@@ -0,0 +1,133 @@
+</template>
+
+<script>
+import Checkmark from '~/assets/icons/checkmark.svg?inline'
```
Nice, I didn't know this was possible! (Might be good to standardise usage between inline SVG imports and the `VIcon` component.)

obulat replied:
It's the [`vue-svg-loader` that is used by the Nuxt `svg-module`](https://github.com/nuxt-community/svg-module#vue-svg-loader). The sad thing about it is I don't think they support Nuxt 3...

sarayourfriend reviewed (changes requested):
Looks really good so far! Just a couple of suggestions, none of which are actually blockers.

The only blocker is a regression in the tooltip for each checkbox. On `main` you can open the tooltip for each license type, but on this branch it doesn't work.

obulat commented on src/components/Filters/FilterChecklist.vue:
```diff
@@ -1,147 +1,101 @@
+        :aria-label="$t('browse-page.aria.license-explanation')"
+        class="appearance-none"
+        type="button"
+        @click.stop="toggleLicenseExplanationVisibility(item.code)"
```
@sarayourfriend , There were two event handlers that got triggered when you clicked the License explanation button: one opened it, and the other closed it right away. Adding `.stop` to the `click` event fixed this.
There is still a problem of the explanation opening in weird places, that I think the `VPopover` could solve.

sarayourfriend replied:
Nice! Yeah using the popover component would be great here! We'd have to add support for the arrow though, I left that out cause there weren't any examples I found in the designs.

I think it would also fix another issue I saw where if you have a license explanation popover open and click on another license explanation button it only closes the existing one but doesn't open the one you clicked on. I think VPopover would fix this too.

sarayourfriend reviewed (approved):
🚀 🌵 🤠 🐄 🌵 🚀

obulat commented on src/components/VCheckbox.vue:
```diff
@@ -0,0 +1,122 @@
+    const onChange = () => {
+      localCheckedState.value = !localCheckedState.value
+      emit('change', {
+        name: inputAttrs.value.name,
```
I like having the properties similar to the native HTML checkbox ones...
